reactioncommerce/reaction
# (fix): Errors when updating default shipping and billing addresses

Resolves #3672
Impact: major
Type: bugfix

## Issue
When changing default shipping and billing addresses, if addresses were updated multiple times in succession, the UI would eventually stop working, causing the user to be stuck with whatever address they were currently locked on.

## Solution
The bulk of the issue seems to stem from our `afterAccountsUpdate` hook, which is used to update the Accounts search record. This hook was running three times every time the `addressBookUpdate` function was run: once for the shipping address, once for the billing address, and once at the end of the function.

The following error is returned when changing the address too much in too short of a span.

```
Exception while invoking method 'accounts/addressBookUpdate' WriteError({"code":11000,"index":0,"errmsg":"E11000 duplicate key error collection: meteor.AccountSearch index: _id_ dup key: { : \"oW372kXYR8Yi3gE2M\" }","op":{"_id":"oW372kXYR8Yi3gE2M","shopId":"J8Bhq3uTtdgwZx3rz","emails":[--------------address info redacted--------------)
    at Function.MongoError.create (/Users/ek/.meteor/packages/npm-mongo/.2.2.34.6do6sz.tll37++os+web.browser+web.cordova/npm/node_modules/mongodb-core/lib/error.js:31:11)
    at toError (/Users/ek/.meteor/packages/npm-mongo/.2.2.34.6do6sz.tll37++os+web.browser+web.cordova/npm/node_modules/mongodb/lib/utils.js:139:22)
    at /Users/ek/.meteor/packages/npm-mongo/.2.2.34.6do6sz.tll37++os+web.browser+web.cordova/npm/node_modules/mongodb/lib/collection.js:668:23
    at handleCallback (/Users/ek/.meteor/packages/npm-mongo/.2.2.34.6do6sz.tll37++os+web.browser+web.cordova/npm/node_modules/mongodb/lib/utils.js:120:56)
    at resultHandler (/Users/ek/.meteor/packages/npm-mongo/.2.2.34.6do6sz.tll37++os+web.browser+web.cordova/npm/node_modules/mongodb/lib/bulk/ordered.js:421:14)
    at /Users/ek/.meteor/packages/npm-mongo/.2.2.34.6do6sz.tll37++os+web.browser+web.cordova/npm/node_modules/mongodb-core/lib/connection/pool.js:469:18
    at _combinedTickCallback (internal/process/next_tick.js:131:7)
    at process._tickDomainCallback (internal/process/next_tick.js:218:9)
```

That error seems to stem from [here](https://github.com/reactioncommerce/reaction/blob/release-1.8.0/imports/plugins/included/search-mongo/server/methods/searchcollections.js#L287-L293). The `_id` is a required field, so if it's being inserted twice - which may happen if the function is called twice (or three times, as stated above) - it will throw the error.

Because of this error, the remainder of the function is never run, which would cause instances where the new address has updated to set `isShippingDefault` and `isBillingDefault` to equal `true`, and never unset the older addresses to be false, so both sets of addresses were set to the defaults.

The fix here seems to be to only run the `afterAccountsUpdate` hook once, at the end of the `addressBookUpdate` function. This will ensure all data is updated as it should be before the hook is run, and not run `afterAccountsUpdate` unnecessarily on every change, instead running it after all the changes are complete.

In addition, I've reworked the `afterAccountsUpdate` hook to force the second parameter to be an object, which contains the user `_id` (which was being passed as a solo string before), as well as an `Array` of fields of the users profile that were updated prior to this hook running.

Because our search only searches `emails`, `firstName`, `lastName`, and `phone`, it does not make sense to re-index the search if a field other than these is updated.

The updated fields `Array` is run against the `Array` of fields used in search, and if a match is found, then the indexing is run.

In addition, I've added `forceIndex` as an option on the updated fields array, so that can be passed at anytime, and is also passed in various instances, such as when a new address is created or an address is deleted.

## Breaking changes
The `afterAccountsUpdate` hook now requires the second parameter to be an object, instead of a string of just the user id. The object still contains `accountId`, and also contains `updatedFields`.

## Testing
1. Log in to your Reaction account
2. Navigate to your Profile page
3. Add multiple addresses to your account
4. Click on the addresses in random orders, making different addresses the default
5. See that you can update your address
6. Repeat above steps in address book of the checkout
